File: backend/python-api/app/service/admin_user_service.py
from __future__ import annotations

import asyncio
import logging

from passlib.hash import bcrypt

from app.domain.admin_action import AdminAction
from app.domain.errors import ERR_CONFLICT, ERR_INVALID_INPUT, DomainError
from app.domain.user import User
from app.dto.admin import AdminCreateUserRequest, AdminUpdateUserRequest, AdminUpdateRoleRequest, UserListParams
from app.i18n.loader import is_language_supported
from app.platform.email import EmailSender
from app.repository.admin_action_repo import AdminActionRepository
from app.repository.user_repo import UserRepository

logger = logging.getLogger(__name__)


class AdminUserService:
    def __init__(
        self,
        users: UserRepository,
        admin_actions: AdminActionRepository,
        email_sender: EmailSender,
    ):
        self._users = users
        self._admin_actions = admin_actions
        self._email = email_sender

    async def list_users(self, params: UserListParams) -> tuple[list[User], int]:
        return await self._users.list_users(params)

    async def get_by_id(self, user_id: int) -> User:
        return await self._users.find_by_id(user_id)

    async def create_user(
        self, req: AdminCreateUserRequest, admin_id: int, admin_email: str, session
    ) -> User:
        existing = await self._users.find_by_email_unscoped(req.email)
        if existing and not existing.is_deleted:
            raise DomainError(ERR_CONFLICT, "User already exists")

        if req.role not in ("user", "admin"):
            raise DomainError(ERR_INVALID_INPUT, "Invalid role")

        lang = req.language.lower() if req.language and is_language_supported(req.language.lower()) else "en"
        password_hash = bcrypt.using(rounds=12).hash(req.password)

        user = User(
            first_name=req.first_name,
            last_name=req.last_name,
            email=req.email,
            password_hash=password_hash,
            role=req.role,
            language=lang,
            verified=req.verified if req.verified is not None else False,
        )
        await self._users.create(user)

        action = AdminAction(
            admin_id=admin_id,
            admin_email=admin_email,
            action="create_user",
            target_user_id=user.id,
            target_email=user.email,
            message=f"Created user {user.email} with role {user.role}",
        )
        await self._admin_actions.create(action)
        await session.commit()
        return user

    async def update_user(
        self, user_id: int, req: AdminUpdateUserRequest, admin_id: int, admin_email: str, session
    ) -> User:
        user = await self._users.find_by_id(user_id)
        old_role = user.role

        if req.first_name is not None:
            user.first_name = req.first_name
        if req.last_name is not None:
            user.last_name = req.last_name
        if req.email is not None:
            user.email = req.email
        if req.password is not None:
            user.password_hash = bcrypt.using(rounds=12).hash(req.password)
        if req.role is not None:
            if req.role not in ("user", "admin"):
                raise DomainError(ERR_INVALID_INPUT, "Invalid role")
            user.role = req.role
        if req.language is not None:
            if is_language_supported(req.language.lower()):
                user.language = req.language.lower()
        if req.country is not None:
            user.country = req.country.upper()[:10]
        if req.address is not None:
            user.address = req.address
        if req.phone is not None:
            user.phone = req.phone
        if req.verified is not None:
            user.verified = req.verified
        if req.two_fa_enabled is not None:
            user.two_fa_enabled = req.two_fa_enabled
            if not req.two_fa_enabled:
                user.two_fa_secret = None

        await self._users.update(user)

        action = AdminAction(
            admin_id=admin_id,
            admin_email=admin_email,
            action="update_user",
            target_user_id=user.id,
            target_email=user.email,
            message=f"Updated user {user.email}",
        )
        await self._admin_actions.create(action)
        await session.commit()

        if req.role and req.role != old_role:
            asyncio.create_task(
                self._email.send_role_change(user.email, user.language or "en", user.full_name, old_role, user.role)
            )

        return user

    async def delete_user(self, user_id: int, admin_id: int, admin_email: str, session) -> None:
        user = await self._users.find_by_id(user_id)

        action = AdminAction(
            admin_id=admin_id,
            admin_email=admin_email,
            action="delete_user",
            target_user_id=user.id,
            target_email=user.email,
            message=f"Deleted user {user.email}",
        )
        await self._admin_actions.create(action)
        await self._users.soft_delete(user.id)
        await session.commit()

        asyncio.create_task(
            self._email.send_account_deleted(user.email, user.language or "en", user.full_name, user.email)
        )

    async def update_role(
        self, user_id: int, req: AdminUpdateRoleRequest, admin_id: int, admin_email: str, session
    ) -> User:
        if req.role not in ("user", "admin"):
            raise DomainError(ERR_INVALID_INPUT, "Invalid role")

        user = await self._users.find_by_id(user_id)
        old_role = user.role
        user.role = req.role
        await self._users.update(user)

        action = AdminAction(
            admin_id=admin_id,
            admin_email=admin_email,
            action="update_role",
            target_user_id=user.id,
            target_email=user.email,
            message=f"Changed role from {old_role} to {req.role}",
        )
        await self._admin_actions.create(action)
        await session.commit()

        asyncio.create_task(
            self._email.send_role_change(user.email, user.language or "en", user.full_name, old_role, user.role)
        )
        return user

    async def stats(self) -> dict:
        total = await self._users.count()
        admin_count = await self._users.count_by_role("admin")
        user_count = await self._users.count_by_role("user")
        return {"total_users": total, "admin_users": admin_count, "regular_users": user_count}

    async def get_actions(self, page: int, limit: int) -> tuple[list[AdminAction], int]:
        return await self._admin_actions.list_actions(page, limit)

File: backend/go-api/internal/service/admin_user_service.go
package service

import (
	"context"
	"fmt"
	"log"
	"strings"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/dto"
	"saas-starter/backend/go-api/internal/platform/email"
	"saas-starter/backend/go-api/internal/repository"
	"saas-starter/backend/go-api/pkg/i18n"

	"golang.org/x/crypto/bcrypt"
)

type AdminUserService struct {
	users    repository.UserRepository
	auditLog repository.AdminActionRepository
	txManager repository.TxManager
	email    email.Sender
}

func NewAdminUserService(
	users repository.UserRepository,
	auditLog repository.AdminActionRepository,
	txManager repository.TxManager,
	emailSender email.Sender,
) *AdminUserService {
	return &AdminUserService{
		users: users, auditLog: auditLog, txManager: txManager, email: emailSender,
	}
}

func (s *AdminUserService) List(ctx context.Context, params dto.UserListParams) ([]domain.User, int64, error) {
	return s.users.List(ctx, params)
}

func (s *AdminUserService) GetByID(ctx context.Context, id uint) (*domain.User, error) {
	return s.users.FindByID(ctx, id)
}

func (s *AdminUserService) Create(ctx context.Context, adminID uint, req dto.AdminCreateUserRequest) (*domain.User, error) {
	existing, err := s.users.FindByEmail(ctx, req.Email)
	if err == nil && existing != nil {
		return nil, domain.NewError(domain.ErrConflict, "user already exists")
	}

	hash, err := bcrypt.GenerateFromPassword([]byte(req.Password), BcryptCost)
	if err != nil {
		return nil, err
	}

	role := req.Role
	if role == "" {
		role = "user"
	}
	validRoles := map[string]bool{"user": true, "admin": true}
	if !validRoles[role] {
		return nil, domain.NewError(domain.ErrInvalidInput, "invalid role")
	}

	lang := normalizeLang(req.Language)

	user := &domain.User{
		FirstName:    req.FirstName,
		LastName:     req.LastName,
		Email:        req.Email,
		PasswordHash: string(hash),
		Role:         role,
		Language:     lang,
	}
	if req.Verified != nil {
		user.Verified = *req.Verified
	}

	if err := s.users.Create(ctx, user); err != nil {
		return nil, err
	}

	s.logAction(ctx, adminID, "create", user, fmt.Sprintf("Created user: %s (%s)", user.FullName(), user.Email))

	return user, nil
}

func (s *AdminUserService) Update(ctx context.Context, adminID, userID uint, req dto.AdminUpdateUserRequest) (*domain.User, error) {
	user, err := s.users.FindByID(ctx, userID)
	if err != nil {
		return nil, err
	}

	oldRole := user.Role
	roleChanged := false

	if req.FirstName != "" {
		user.FirstName = req.FirstName
	}
	if req.LastName != "" {
		user.LastName = req.LastName
	}
	if req.Email != "" {
		user.Email = req.Email
	}
	if req.Password != "" {
		hash, err := bcrypt.GenerateFromPassword([]byte(req.Password), BcryptCost)
		if err != nil {
			return nil, err
		}
		user.PasswordHash = string(hash)
	}
	if req.Role != "" {
		validRoles := map[string]bool{"user": true, "admin": true}
		if !validRoles[req.Role] {
			return nil, domain.NewError(domain.ErrInvalidInput, "invalid role")
		}
		if user.Role != req.Role {
			roleChanged = true
			user.Role = req.Role
		}
	}
	if req.Language != "" {
		lang := req.Language
		if len(lang) > 2 {
			lang = lang[:2]
		}
		lang = strings.ToLower(lang)
		if i18n.IsLanguageSupported(lang) {
			user.Language = lang
		}
	}
	if req.Country != "" {
		c := strings.ToUpper(req.Country)
		if len(c) > 10 {
			c = c[:10]
		}
		user.Country = c
	} else {
		user.Country = ""
	}

	user.Address = req.Address
	user.Phone = req.Phone

	if req.Verified != nil {
		user.Verified = *req.Verified
	}
	if req.TwoFAEnabled != nil {
		user.TwoFAEnabled = *req.TwoFAEnabled
		if !*req.TwoFAEnabled && user.TwoFASecret != "" {
			user.TwoFASecret = ""
		}
	}

	if err := s.users.Update(ctx, user); err != nil {
		return nil, err
	}

	s.logAction(ctx, adminID, "update", user, fmt.Sprintf("Updated user: %s (%s)", user.FullName(), user.Email))

	if roleChanged {
		s.logAction(ctx, adminID, "role_change", user, fmt.Sprintf("Changed role from '%s' to '%s' for %s", oldRole, user.Role, user.Email))
		go func() {
			if err := s.email.SendRoleChange(user.Email, user.PreferredLang("en"), user.FullName(), oldRole, user.Role); err != nil {
				log.Printf("Failed to send role change notification to %s: %v", user.Email, err)
			}
		}()
	}

	return user, nil
}

func (s *AdminUserService) Delete(ctx context.Context, adminID, userID uint) error {
	user, err := s.users.FindByID(ctx, userID)
	if err != nil {
		return err
	}

	err = s.txManager.WithTx(ctx, func(txCtx context.Context) error {
		s.logAction(txCtx, adminID, "delete", user, fmt.Sprintf("Deleted user: %s (%s)", user.FullName(), user.Email))
		return s.users.SoftDelete(txCtx, userID)
	})
	if err != nil {
		return err
	}

	go func() {
		if sendErr := s.email.SendAccountDeleted(user.Email, user.PreferredLang("en"), user.FullName(), user.Email); sendErr != nil {
			log.Printf("Failed to send deletion notification to %s: %v", user.Email, sendErr)
		}
	}()

	return nil
}

func (s *AdminUserService) UpdateRole(ctx context.Context, adminID, userID uint, role string) (*domain.User, error) {
	user, err := s.users.FindByID(ctx, userID)
	if err != nil {
		return nil, err
	}

	oldRole := user.Role
	user.Role = role

	if err := s.users.Update(ctx, user); err != nil {
		return nil, err
	}

	s.logAction(ctx, adminID, "role_change", user, fmt.Sprintf("Changed role from '%s' to '%s' for %s", oldRole, role, user.Email))

	go func() {
		if err := s.email.SendRoleChange(user.Email, user.PreferredLang("en"), user.FullName(), oldRole, role); err != nil {
			log.Printf("Failed to send role change notification to %s: %v", user.Email, err)
		}
	}()

	return user, nil
}

func (s *AdminUserService) Stats(ctx context.Context) (map[string]int64, error) {
	total, _ := s.users.Count(ctx)
	admins, _ := s.users.CountByRole(ctx, "admin")
	regular, _ := s.users.CountByRole(ctx, "user")

	return map[string]int64{
		"total_users":   total,
		"admin_users":   admins,
		"regular_users": regular,
	}, nil
}

func (s *AdminUserService) GetActions(ctx context.Context, page, limit int) ([]domain.AdminAction, int64, error) {
	return s.auditLog.List(ctx, page, limit)
}

func (s *AdminUserService) logAction(ctx context.Context, adminID uint, action string, target *domain.User, message string) {
	admin, _ := s.users.FindByID(ctx, adminID)
	adminEmail := "unknown"
	if admin != nil && admin.Email != "" {
		adminEmail = admin.Email
	}

	entry := &domain.AdminAction{
		AdminID:      adminID,
		AdminEmail:   adminEmail,
		Action:       action,
		TargetUserID: target.ID,
		TargetEmail:  target.Email,
		Message:      message,
	}
	if err := s.auditLog.Create(ctx, entry); err != nil {
		log.Printf("Failed to log admin action: %v", err)
	}
}
